Add h7 power states

benediktibk opened the pull request:
Implement power states for STM32H735 and test them in a sample project on the discovery kit.

erwango commented on soc/arm/st_stm32/stm32h7/power_m7.c:
```diff
@@ -0,0 +1,57 @@
+		defined(CONFIG_POWER_SUPPLY_SMPS_2V5_SUPPLIES_EXT))
+#error Unsupported configuration: Selected SoC do not support SMPS
+#endif
+#if defined(CONFIG_POWER_SUPPLY_DIRECT_SMPS)
+	LL_PWR_ConfigSupply(LL_PWR_DIRECT_SMPS_SUPPLY);
+#elif defined(CONFIG_POWER_SUPPLY_SMPS_1V8_SUPPLIES_LDO)
+	LL_PWR_ConfigSupply(LL_PWR_SMPS_1V8_SUPPLIES_LDO);
+#elif defined(CONFIG_POWER_SUPPLY_SMPS_2V5_SUPPLIES_LDO)
```
While you're at it, would you mind using `IS_ENABLED` instead of `#ifdef's` for readability ?
```suggestion
	if (IS_ENABLED(CONFIG_POWER_SUPPLY_DIRECT_SMPS)) {
		LL_PWR_ConfigSupply(LL_PWR_DIRECT_SMPS_SUPPLY);
	} else if (IS_ENABLED(CONFIG_POWER_SUPPLY_SMPS_1V8_SUPPLIES_LDO)) {
		LL_PWR_ConfigSupply(LL_PWR_SMPS_1V8_SUPPLIES_LDO);
	.....
```
Would be nice...

benediktibk replied:
done

benediktibk replied:
Unfortunately I had to switch back to the #if defined approach, as otherwise the code would not build for some platforms. For some platforms the symbol LL_PWR_DIRECT_SMPS_SUPPLY for instance might not be defined.

erwango commented on soc/arm/st_stm32/stm32h7/power_m7.c:
```diff
@@ -0,0 +1,57 @@
+	return 0;
+}
+
+SYS_INIT(stm32_power_init, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
Is `POST_KERNEL` suitable ?
At that step GPIO and UART are already up and running

benediktibk replied:
Moved it to PRE_KERNEL_1, as this seems to be more suitable. In the typical professional software developer style I originally just copied this from STM32L4. For STM32G0 it is the same. I suppose they should be changed as well, but I'm kinda hesitant to do so as I do not have these devices available to test the changes.

erwango replied:
> I suppose they should be changed as well, but I'm kinda hesitant to do so as I do not have these devices available to test the changes.

This should be done as part as another PR anyway. Raising a dedicated issue would be fine.

benediktibk replied:
Opened a separate issue: #41442

erwango commented on soc/arm/st_stm32/stm32h7/power_m7.c:
```diff
@@ -48,8 +48,15 @@ static int stm32_power_init(const struct device *dev)
 	LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE1);
+
+	/*
+	 * At this point it should actually only be necessary to wait for ACTVOSRDY during startup.
```
Might use some rewording here: "it should actually only be necessary to wait for ACTVOSRDY".
That's not what we're doing as we're waiting for both VOS and ACTVOSRDY.

erwango commented on soc/arm/st_stm32/stm32h7/power_m7.c:
```diff
@@ -48,8 +48,15 @@ static int stm32_power_init(const struct device *dev)
 	LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE1);
+
+	/*
+	 * At this point it should actually only be necessary to wait for ACTVOSRDY during startup.
+	 * Anyway, whenever the voltage scaling is changed one you should also wait for VOSRDY.
```
Comment is a bit confusing even if I finally got it in the end.
Could you have a true to make it clearer ?

benediktibk replied:
I agree, it is confusing. Hopefully(?) done.

benediktibk replied:
Adapted the comment slightly.

erwango commented on soc/arm/st_stm32/stm32h7/CMakeLists.txt:
```diff
@@ -4,6 +4,7 @@ zephyr_include_directories(${ZEPHYR_BASE}/drivers)
 
 zephyr_sources_ifdef(CONFIG_CPU_CORTEX_M7 soc_m7.c)
 zephyr_sources_ifdef(CONFIG_CPU_CORTEX_M7 power_m7.c)
+zephyr_sources_ifdef(CONFIG_SOC_STM32H735XX power_stm32h735xx.c)
```
Are there some reasons it can't be more widely used ?

benediktibk replied:
So far I based my work upon the reference manual for H723, H733, H725, H735 and H730. Therefore I will have to check if the same procedures also apply to H743, H745, H747, H750 and H753.
As soon as I did so I will update this line.

erwango replied:
Thanks.
It's possible that there are some differences between these SoC.
Maybe we'll need to introduce some device tree configuration to be able to group them by specifics

benediktibk replied:
I went through the other reference manuals, and of course there is a difference. H743, H745, H747, H750 and H753 require an additional step for changing into VOS0. For these chips this switch is only allowed from VOS1 and with the ODEN bit in SYSCFG_PWRCR set. Hence, only the voltage scaling is different.

My idea for this would be to make a separate file for the voltage scaling which gets included to the build by a new config param:

`zephyr_sources_ifdef(CONFIG_SOC_STM32H7_VOS_ODEN power_voltage_overdrive.c)
zephyr_sources_ifdef(CONFIG_SOC_STM32H7_VOS_WITHOUT_ODEN power_voltage.c)`

I would set these flags automatically based upon the chip selection.

Does this sound reasonable to you?

benediktibk replied:
I switched to a slightly different way using a menu choice. Hopefully more elegant.

erwango commented on soc/arm/st_stm32/stm32h7/power_stm32h735xx.c:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	switch (info.substate_id) {
+	case 1: /* system mode SLEEP with VOS0*/
+		k_cpu_idle();
+		break;
+	case 2: /* system mode SLEEP with VOS1*/
+		LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE1);
+
+		while (LL_PWR_IsActiveFlag_VOS() == 0) {
+		}
+
+		k_cpu_idle();
+		break;
+	case 3: /* system mode SLEEP with VOS2*/
+		LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE2);
+
+		while (LL_PWR_IsActiveFlag_VOS() == 0) {
+		}
+
+		k_cpu_idle();
+		break;
+	case 4: /* system mode SLEEP with VOS3*/
+		LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE3);
+
+		while (LL_PWR_IsActiveFlag_VOS() == 0) {
+		}
+
+		k_cpu_idle();
+		break;
```
Seems like these states should be sub-states of `PM_STATE_ACTIVE/PM_STATE_RUNTIME_IDLE` rather than `PM_STATE_SUSPEND_TO_IDLE`.
Cf states definitions is `include/pm/state.h`

benediktibk replied:
Moved everything to PM_STATE_RUNTIME_IDLE

erwango replied:
Well, the cases 5 to 10 were rather correct probably (and should be kept to PM_STATE_SUSPEND_TO_IDLE).
Cf how to handle several states and substates in `soc/arm/st_stm32/stm32u5/power.c`.

Note that, from what I know, the states are up to now used with power states
and this is the first time someone attempts to perform voltage scaling, so I can't say for sure this will work
as you expect.
Please note that another way to achieve this would be to un-correlate both pm states and VOS states,
by using separate functions.

Maybe @ceolin or @gmarull have recommendations here.

benediktibk replied:
done, so far

erwango commented on samples/subsys/pm/power_state/README.rst:
```diff
@@ -0,0 +1,37 @@
+.. _power_state:
```
Move this to samples/boards/stm32/power_mgmt and make it H7 specific.

benediktibk replied:
done

erwango commented on samples/subsys/pm/power_state/src/main.c:
```diff
@@ -0,0 +1,36 @@
+
+void main(void)
+{
+	GPIO_InitTypeDef GPIO_InitStruct;
```
I don't have a zephyr alternative to propose to the use of `GPIO_AF0_C1DSLEEP`.
Though, it would be better to demonstrate PM using zephyr API.
Can't `samples/boards/stm32/power_mgmt/blinky` be reused here?

benediktibk replied:
With this example I wanted to check if I am correctly switching into sleep and deep sleep, hence the specific usage of this alternate functions. Technically we could also throw the whole example away again, I basically used it only for testing. Just thought it might be useful for somebody else as well, for instance if one wants to implement other power states.
But I get the issue, it is not so nice to use HAL directly.

erwango replied:
Let's keep it for now if useful for testing. It's valuable for others to have something to test during review.
Point is that we'll need to rework/remove it before merge, but we're not there yet.

benediktibk replied:
Removed the example, as I do not need it anymore.

erwango commented on dts/arm/st/h7/stm32h735.dtsi:
```diff
@@ -49,6 +49,66 @@
+			interrupts = <138 1>;
+			interrupt-names = "wakeup";
+			status = "disabled";
+			label = "LPTIM_2";
+		};
+
+		lptim3: timers@58002800 {
+			compatible = "st,stm32-lptim";
+			clocks = <&rcc STM32_CLOCK_BUS_APB4 0x00000400>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x58002800 0x400>;
+			interrupts = <139 1>;
+			interrupt-names = "wakeup";
+			status = "disabled";
+			label = "LPTIM_3";
+		};
+
+		lptim4: timers@58002c00 {
+			compatible = "st,stm32-lptim";
+			clocks = <&rcc STM32_CLOCK_BUS_APB4 0x00000800>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x58002c00 0x400>;
+			interrupts = <140 1>;
+			interrupt-names = "wakeup";
+			status = "disabled";
+			label = "LPTIM_4";
+		};
+
+		lptim5: timers@58003000 {
+			compatible = "st,stm32-lptim";
+			clocks = <&rcc STM32_CLOCK_BUS_APB4 0x00001000>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x58003000 0x400>;
+			interrupts = <141 1>;
+			interrupt-names = "wakeup";
+			status = "disabled";
+			label = "LPTIM_5";
```
LPTIM driver implementation doesn't support multiples instances today.
Let's remove instances 2 to 5 for now in order to avoid confusion.

benediktibk replied:
done

erwango commented on boards/arm/stm32h735g_disco/stm32h735g_disco_defconfig:
```diff
@@ -26,3 +26,9 @@ CONFIG_PINCTRL=y
 CONFIG_POWER_SUPPLY_DIRECT_SMPS=y
+
+# use LPTIM as clock instead of SYSTICK
+CONFIG_CORTEX_M_SYSTICK=n
```
Enabling lptim1 in device tree + CONFIG_PM=y should be sufficient to perform this setting.
Please check how this is achived on other series such as L4.
I don't think there is much to do to get it working similarly on on H7 .

benediktibk replied:
done

erwango commented on boards/arm/stm32h735g_disco/stm32h735g_disco_defconfig:
```diff
@@ -26,3 +26,9 @@ CONFIG_PINCTRL=y
+# use LPTIM as clock instead of SYSTICK
+CONFIG_CORTEX_M_SYSTICK=n
+
+# activate power management
+CONFIG_PM=y
```
CONFIG_PM should not be enabled as part of default board configuration.
This configuartion is application responsibility and should be done in samples.

benediktibk replied:
done

erwango commented on soc/arm/st_stm32/stm32h7/soc_m7.c:
```diff
@@ -63,12 +63,11 @@ static int stm32h7_init(const struct device *arg)
-	if (!(SCB->CCR & SCB_CCR_DC_Msk)) {
-		SCB_EnableDCache();
+	if (!IS_ENABLED(CONFIG_NOCACHE_MEMORY)) {
+		if (!(SCB->CCR & SCB_CCR_DC_Msk)) {
+			SCB_EnableDCache();
+		}
 	}
-
-#endif /* CONFIG_NOCACHE_MEMORY */
```
I agree this code has to be fixed. But doing it has impacts and should be done in a dedicated PR, carefully taking all the impacts into account (use of DMA for instance).

benediktibk replied:
I can easily revert this changes, they were just additional ifdefs which I replaced as requested with IS_ENABLED. As I see it this should result in the same logic as previously?

erwango replied:
My bad, I missed the IS_ENABLED and understood you purely removed the `CONFIG_NOCACHE_MEMORY` condition.
So it's not bad, but to avoid further confusion, it might be better to revert.
(And limit your changes to PM in this PR)

benediktibk replied:
done

erwango commented:
@benediktibk Thanks for this work. I've performed a first round of comments.
I hope this will help guiding you in this task. I will likely resume review in January now.

erwango commented on boards/arm/stm32h735g_disco/Kconfig.defconfig:
```diff
@@ -12,5 +12,7 @@ config SPI_STM32_INTERRUPT
 	default y
 	depends on SPI
 
+config STM32_LPTIM_TIMER
+	default y if PM
```
Move to soc/arm/st_stm32/stm32h7/Kconfig.defconfig.series

benediktibk replied:
done

benediktibk committed:
soc: arm: stm32h7: implement power states

Implement power states for H7 series SOCs and add an example
to test the wakeup from soft off via RTC.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

erwango commented on boards/arm/stm32h735g_disco/Kconfig.defconfig:
```diff
@@ -12,5 +12,4 @@ config SPI_STM32_INTERRUPT
 	default y
 	depends on SPI
 
-
```
Please remove this

erwango commented on boards/arm/stm32h735g_disco/stm32h735g_disco_defconfig:
```diff
@@ -26,3 +26,6 @@ CONFIG_CLOCK_CONTROL=y
 
 # enable pin controller
 CONFIG_PINCTRL=y
+
+# select power supply
+CONFIG_POWER_SUPPLY_DIRECT_SMPS=y
```
Let's make SMPS configuration a separate PR.

benediktibk replied:
https://github.com/zephyrproject-rtos/zephyr/pull/41868

erwango commented on dts/arm/st/h7/stm32h735.dtsi:
```diff
@@ -49,6 +49,18 @@
 			health-test-magic = <0x17590abc>;
 			health-test-config = <0xaa74>;
 		};
+
+		lptim1: timers@40002400 {
+			compatible = "st,stm32-lptim";
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00000200>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40002400 0x400>;
+			interrupts = <93 1>;
+			interrupt-names = "wakeup";
+			status = "disabled";
+			label = "LPTIM_1";
+		};
```
Let's make LPTIM support on H7 a separate PR

benediktibk replied:
https://github.com/zephyrproject-rtos/zephyr/pull/41867

erwango commented on soc/arm/st_stm32/stm32h7/Kconfig.defconfig.stm32h723xx:
```diff
@@ -11,4 +12,9 @@ config SOC
 config NUM_IRQS
 	default 163
 
+choice VOS0_SWITCH_METHOD_CHOICE
+	default VOS0_SWITCH_METHOD_DIRECT
+
+endchoice
+
```
Please make changes around VOS switch a separate PR.

benediktibk replied:
https://github.com/zephyrproject-rtos/zephyr/pull/41869

erwango reviewed (changes requested):
@benediktibk Thanks for this work.
This is a rather big change that introduce multiple parts, some could be rather fast to review and integrate while I expect some others to be require more thoughts and reviews.
In order to be able to progress faster I propose to split this change into multiple PRs
- LPTIM support on H7
- SMPS support on H735 disco 
- VOS Switch
- PM support on H7
In these PR, please try to split in commits when possible.
